### PR TITLE
Implement actor sketch

### LINKED ISSUE #1: Implement the crate by extracting code from `pd`
The `tower::Service` trait is very versatile, and allows writing a wide variety of composable middleware.  However, one gap in existing middleware is bridging between the request/response-oriented `Service` interface, which is primarily designed around the idea that requests are processed independently and asynchronously, and an actor-like model where a stateful actor processes a queue of incoming messages.

Tower does use an actor model in the form of the `Buffer` middleware for mediating access to a shared `Service`, by moving the `Service` into a worker task that pulls request messages from a `mpsc`, and presenting a `Buffer` handle that wraps the sender side of the `mpsc`.  This is almost what we want, except that the `Buffer` can only wrap a `Service` implementation, and the `Service` trait isn't quite what we want: while the `Service` trait's `call` method allows _synchronous_ mutation of state (via `&mut self`), it doesn't work well with _asynchronous_ mutation of state (since all asynchronous work has to be done in the response future, which can't have access to the `Service` state without additional sharing and synchronization setup).

As a result, in `pd`, we ended up vendoring and forking the `Buffer` middleware, and using the forked code to wrap hardcoded worker tasks for the consensus and mempool services we present to `tower-abci`.  It would be much better, though, if we could make that interface useful to other users of `tower-abci`.

To accomplish this goal, we should extract and generalize the code in `pd` (found [here](https://github.com/penumbra-zone/penumbra/blob/main/pd/src/consensus/service.rs#L17-L20) and [here](https://github.com/penumbra-zone/penumbra/blob/main/pd/src/mempool/service.rs#L23-L26)) into a standalone `tower-actor` crate.

The externally-facing API should be minimal and look roughly like this (it could and should be split up into different files, this is just the parts that would be re-exported from the crate root):

```rust
pub struct Message<R, S, E> {
    pub req: R,
    pub rsp_sender: oneshot::Sender<Result<S, E>>,
    pub span: tracing::Span,
}

pub struct Actor<R, S, E> {
    queue: tokio_util::sync::PollSender<Message<R, S, E>>
}

impl<R, S, E> Actor<R, S, E> {
    pub fn new<F, W>(bound: usize, f: F) -> Self
    where
        F: FnOnce(mpsc::Receiver<Message<R, S, E>>) -> W,
        W: Future<Output = Result<(), E>> + Send + 'static,
    {
        let (queue_tx, queue_rx) = mpsc::channel(bound);

        tokio::task::Builder::new()
            .name("tower-actor-worker")
            .spawn(f(queue_rx))
            .expect("failed to spawn worker");

        Self {
            queue: PollSender::new(queue_tx),
        }
    }
}

pub struct ActorFuture<S, E> {
    inner: oneshot::Receiver<Result<S, E>>,
}

impl<S, E> Future for ActorFuture<S, E> {
    type Output = Result<S, E>;
    // ... (wrapper around oneshot future impl) ...
}

impl<R, S, E> Service<R> for Actor<R, S, E>
where
    E: From<tokio_util::sync::PollSendError>
{
    type Response = S;
    type Error = E;
    type Future = ActorFuture<S, E>;
    
    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.queue.poll_reserve(cx).map_err(Into::into)
    }

    fn call(&mut self, req: R) -> Self::Future {
        // construct and enqueue Message, return a wrapped oneshot receiver
    }
}
```

Some notes on the API:

- Including the `Span` in the `Message` allows propagating semantic context to the worker task.
- Because the `Message` returns the response via a `oneshot`, in principle the worker can choose how much work to do sequentially, either by completely processing the entire request and then sending it, or partially processing it and then spawning the rest as another task.
- Returning an opaque `ActorFuture` instead of a oneshot receiver is slightly cleaner; not show in the sketch is that it will need to use `#[pin_project]` [like so](https://github.com/penumbra-zone/penumbra/blob/main/storage/src/future.rs#L23-L35)
- The error handling story here is not 100% optimal, we bound the worker `W` as `Future<Output<Result<(), E>>>` but don't do anything to propagate any error that occurs into the `Actor` handles.  I think this is fine for a first pass, but in the future we could do something like `tower::Buffer`'s error slot design.
- Instead of just having `Actor::new`, we could have `Actor::new_named(name: &str, bound: usize, f: F)` that would pass `name` to the `tokio::task::Builder`; then `Actor::new` can wrap `Actor::new_named("tower-actor-worker", bound, f)`.

### PR DESCRIPTION
**Pull request by mikayla-maki**
Fixes https://github.com/penumbra-zone/tower-actor/issues/1

TODO:

- [x] Figure out good tests for each of the error conditions
- [x] Double check that exports work correctly.
- [x] Get a pass from the penumbra dev team
- [x] Document the library!

OUT OF DATE COMMENTARY:

Modeling the errors in this crate proved to be a bit trickier than I expected. The original code uses `anyhow` + `Box<dyn Error + Send + Sync + 'static>` to erase all error information but this didn't feel like the right approach for a library. I ended up going with [sled's](http://sled.rs/errors.html) approach of returning a `Result<Result<S, E>, ActorError>` to capture the difference between unrecoverable queue errors and theoretically recoverable consumer errors. This also required me to make the actual ActorFuture itself an enum, to capture the case where the queue was closed between `poll_ready()` and `call()`. Curious to hear what y'all think :)

### CODE REVIEW

**Comment by mikayla-maki** [comment]
The error condition referenced above, which required the actor future to become an enum, turned out to be impossible. The documentation of [`PollSender::close()`](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.PollSender.html#method.close) guarantees that you can still send on a queue that has been dropped / `.closed()` / panicked until all of it's permits have been used up. As such, I was able to simplify the error handling significantly and retain the design as shown in the sketch. I added a couple tests for some common cases that I could think of, honestly there's not that much actual code being executed here. Most of the effort has been in setting up the types / interfaces.

**Comment by mikayla-maki** [comment]
Added some compile flags to ensure that projects which include tower_actor but don't use `tokio_unstable` get a useful error message. Also, made sure that all of the re-exports look nice.

**Comment by hdevalence** [comment]
On the error handling front, the reason to prefer a boxed error rather than nesting multiple concrete error type is that it's the only practical way to achieve composability: if you have a single Result<Rsp, BoxError>, the caller code doesn't need to know the details of the whole layer stack, but they can still use downcasting to handle specific error types if they want. By contrast, nested error types reify the layer stack in the response type, so that changing it later (eg rearranging the order of layers) breaks caller code.

File: src/lib.rs [hunk 1]
```diff
@@ -1,14 +1,179 @@
...
+
+#[cfg(tokio_unstable)]
+mod tower_actor {
+    pub(crate) mod actor;
+    pub(crate) mod future;
+    pub(crate) mod message;
+
+    pub use actor::*;
+    pub use future::*;
+    pub use message::*;
```
Comment (hdevalence) [inline src/lib.rs]: The only tokio_unstable feature used here is the named task used to spawn the actor, right?

I think it'd be cleaner to make the crate compile unconditionally, and limit the cfg-gating to the spawn call, using the cfg to switch between either calling spawn directly and ignoring the task name, or using the unstable Builder API if it's available. (Maybe we'll get an unused variable warning in the unavailable case, I think that doesn't matter too much)

File: src/tower_actor/actor.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    fn call(&mut self, req: R) -> Self::Future {
+        // Due to the permit system that PollSender uses, we can always send on the queue if
+        // Poll reserve succeeded.
+        // See: https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.close
+        //
+        // Since the Service contract requires that `poll_ready()` pass
+        // before calling `call()`, we can safely proceed without checking that the queue isn't closed.
+        debug_assert!(!self.queue.is_closed());
+
+        let span = req.create_span();
```
Comment (hdevalence) [inline src/tower_actor/actor.rs]: I think this is a holdover from the pd code, which hooked custom span creation into the service definitions. But in a generic context I don't think it's appropriate for the service to create custom spans as well as doing the actor part — we should just use Span::current() here, and if the caller wants to do custom spam handling, they should do that by composing in an additional tower layer. That way, we only do one thing, and let the user compose behavior as desired.

**Review by hdevalence — CHANGES_REQUESTED** [review]
This looks great so far, thanks so much! I only had two minor comments.

**Comment by mikayla-maki** [comment]
Changes implemented!

**Comment by mikayla-maki** [comment]
While reviewing the code, I also noticed that the type of the actor constructor callback was a little funny, in the sketch it returns a `Result<(), MessageErrorType>`, but the join handle carrying that return is immediately dropped in the constructors. I've switched it to returning the unit type to more clearly express this. 

That said, I think it could be cool if the Actor was itself a future that could be `.await`to get at that return type. But to do that I think we'd need to stuff some types in somewhere to express 'what is this actor's result and error' and then we'd probably have to make an 'actor reference' that holds the poll sender and can be cloned around. Not too hard, but a big change from the existing API. Also I should probably go have a weekend at some point 😅

**Comment by hdevalence** [comment]
Ah, yeah. I think the writeup I did was unclear. The idea behind having the actor future return an error is so that at some point, we can extend the implementation to have a shared "error slot" (like in the Buffer implementation) like Arc<Mutex<Option<BoxError>>> or something, then

- wrap the actor future in a shim that checks if it exited with an error and sets the error slot
- hook the poll_ready implementation to check the error slot and fail the Service if the Actor failed (IIRC there may be some complexity here about needing to clone the error? maybe it has to be Option<Arc<BoxError>> or something so that multiple Services can share the internal error?)

I think we want the future the user passed in to resolve to a Result, even if we currently discard the error, so that later, we can improve the implementation to handle errors better without having to change the external API.

For now I think it's fine to discard the error as an implementation shortcoming, and plan to fix it up later.

**Comment by mikayla-maki** [comment]
Ahhh, I see. I did read the note about that but I didn’t quite put all the pieces together. I’ll look into the buffer stuff and maybe add an error slot implementation later. Thanks for the context :D

**Comment by mikayla-maki** [comment]
Looked up the buffer shim implementation, and it seemed like more effort than this Sunday is calling for. Restored the results as part of a planned future expansion and hopefully this should be good to go now :)

**Pull request merged**